### PR TITLE
EndpointRegistry fixes

### LINKED ISSUE #1845: Unexpected(?) exception in EndpointRegistry.RegisterReadOnlyEndpoint()
Hi, 

I'm using Akka.remote(1.0.6 from nuget) to build pub-sub like service.
When one of pub-sub member restarts its actorsystem while the other is running, using 
following code 

```
  var task = Instance.Terminate(); // Instance of  'ActorSystem' 
  task.Wait();
  Instance = null;
```

Sometims(not always I think) I got following error log in the other side(not the one restarted)

```
2016/04/02 09:29:07.303    11 ERROR Akka.Actor.OneForOneStrategy  동일한 키를 사용하는 항목이 이미 추가되었습니다. System.ArgumentException: 동일한 키를 사용하는 항목이 이미 추가되었습니다.
   위치: System.ThrowHelper.ThrowArgumentException(ExceptionResource resource)
   위치: System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
   위치: Akka.Remote.EndpointRegistry.RegisterReadOnlyEndpoint(Address address, IActorRef endpoint, Int32 uid)
   위치: Akka.Remote.EndpointManager.AcceptPendingReader(IActorRef takingOverFrom)
   위치: Akka.Remote.EndpointManager.<Accepting>b__42_21(Terminated terminated)
   위치: lambda_method(Closure , Object , Action`1 , Action`1 , Action`1 , Action`1 , Action`1 , Action`1 , Action`1 , Action`1 , Action`1 , Action`1 , Action`1 )
   위치: Akka.Tools.MatchHandler.PartialHandlerArgumentsCapture`12.Handle(T value)
   위치: Akka.Actor.ReceiveActor.ExecutePartialMessageHandler(Object message, PartialAction`1 partialAction)
   위치: Akka.Actor.ReceiveActor.<>c__DisplayClass11_0.<Become>b__0(Object m)
   위치: Akka.Actor.ActorCell.<>c__DisplayClass109_0.<Akka.Actor.IUntypedActorContext.Become>b__0(Object m)
   위치: Akka.Actor.ActorBase.AroundReceive(Receive receive, Object message)
   위치: Akka.Actor.ActorCell.ReceiveMessage(Object message)
   위치: Akka.Actor.ActorCell.ReceivedTerminated(Terminated t)
   위치: Akka.Actor.ActorCell.AutoReceiveMessage(Envelope envelope)
   위치: Akka.Actor.ActorCell.Invoke(Envelope envelope)
```

Sorry for the Korean in log, anyway it reads "You are trying to add a new item whose key is already in dictionary" or something.  

Due to my lack of knowledge on EndpointManager, I only have to say I'm feeling like somethings go wrong. sorry for that. but Is this really ok? 

https://github.com/akkadotnet/akka.net/blob/dev/src/core/Akka.Remote/EndpointRegistry.cs#L58

After this happens,  the other actorsystem(not the one who restarted) looks like missing  detectection of  some remote actors(in the actorsystem who restarted)'s death watch.

Any hint or advice will be appreciated.

### PR DESCRIPTION
**Pull request by Aaronontheweb**
Fixed numerous issues with the `EndpointRegistry` and added specs to cover - should resolve #1845 and others.

### CODE REVIEW

File: src/Akka.sln [hunk 1]
```diff
@@ -193,9 +193,6 @@ EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "PersistenceBenchmark", "benchmark\PersistenceBenchmark\PersistenceBenchmark.csproj", "{39E6F51F-FA1E-4C62-B8F8-19065DE6D55D}"
 EndProject
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Solution Items", "Solution Items", "{A59BAE84-70E2-46A0-9E26-7413C103E2D7}"
-	ProjectSection(SolutionItems) = preProject
```
Comment (Aaronontheweb) [inline src/Akka.sln]: No idea what this was for. Deleted it.

File: src/core/Akka.Remote.Tests/EndpointRegistrySpec.cs [hunk 2]
```diff
@@ -148,6 +148,71 @@ public void EndpointRegistry_should_be_able_to_register_Quarantined_policy_for_a
             Assert.Equal(42, reg.WritableEndpointWithPolicyFor(address1).AsInstanceOf<EndpointManager.Quarantined>().Uid);
             Assert.Equal(deadline, reg.WritableEndpointWithPolicyFor(address1).AsInstanceOf<EndpointManager.Quarantined>().Deadline);
         }
+
+        [Fact]
```
Comment (Aaronontheweb) [inline src/core/Akka.Remote.Tests/EndpointRegistrySpec.cs]: Specs to cover the original case reported in #1845 and other cases where bad behavior could occur.

File: src/core/Akka.Remote/EndpointManager.cs [hunk 3]
```diff
@@ -628,6 +628,7 @@ protected void Accepting()
                         _transportMapping[send.Recipient.LocalAddressToUse], _settings, writing: true,
                         handleOption: null, refuseUid: refuseUid), uid: null, refuseUid: refuseUid);
 
+                // pattern match won't throw a NullReferenceException if one is returned by WritableEndpointWithPolicyFor
```
Comment (Aaronontheweb) [inline src/core/Akka.Remote/EndpointManager.cs]: Added a comment here just to clarify - `_endpoints.WritableEndpointWithPolicyFor` can return `null`, but `PatterMatch` will not throw and will route the call to the `Default` match below, which is the correct behavior.

File: src/core/Akka.Remote/EndpointRegistry.cs [hunk 4]
```diff
@@ -37,28 +37,29 @@ public IActorRef RegisterWritableEndpoint(Address address, IActorRef endpoint, i
                 throw new ArgumentException("Attempting to overwrite existing endpoint " + e + " with " + endpoint);
             }
 
-            _addressToWritable.AddOrSet(address, new EndpointManager.Pass(endpoint, uid, refuseUid));
-            _writableToAddress.AddOrSet(endpoint, address);
+            _addressToWritable[address] = new EndpointManager.Pass(endpoint, uid, refuseUid);
```
Comment (Aaronontheweb) [inline src/core/Akka.Remote/EndpointRegistry.cs]: Replacing `AddOrSet` with the built-in operators.

File: src/core/Akka.Remote/EndpointRegistry.cs [hunk 5]
```diff
@@ -37,28 +37,29 @@ public IActorRef RegisterWritableEndpoint(Address address, IActorRef endpoint, i
...
             return endpoint;
         }
 
-        public void RegisterWritableEndpointUid(IActorRef writer, int uid)
+        public void RegisterWritableEndpointUid(Address remoteAddress, int uid)
         {
-            var address = _writableToAddress[writer];
-            if (_addressToWritable[address] is EndpointManager.Pass)
+            EndpointManager.EndpointPolicy existing;
+            if (_addressToWritable.TryGetValue(remoteAddress, out existing))
```
Comment (Aaronontheweb) [inline src/core/Akka.Remote/EndpointRegistry.cs]: Made this code more error-proof.

File: src/core/Akka.Remote/EndpointRegistry.cs [hunk 6]
```diff
@@ -37,28 +37,29 @@ public IActorRef RegisterWritableEndpoint(Address address, IActorRef endpoint, i
...
-
-            // if the policy is not Pass, then the GotUid might have lost the race with some failure
-
         }
 
         public IActorRef RegisterReadOnlyEndpoint(Address address, IActorRef endpoint, int uid)
         {
-            _addressToReadonly.Add(address, Tuple.Create(endpoint, uid));
-            _readonlyToAddress.Add(endpoint, address);
+            _addressToReadonly[address] = Tuple.Create(endpoint, uid);
```
Comment (Aaronontheweb) [inline src/core/Akka.Remote/EndpointRegistry.cs]: Replacing `AddOrSet` with the built-in operators.

File: src/core/Akka.Remote/EndpointRegistry.cs [hunk 7]
```diff
@@ -67,8 +68,9 @@ public void UnregisterEndpoint(IActorRef endpoint)
             if (IsWritable(endpoint))
             {
                 var address = _writableToAddress[endpoint];
-                var policy = _addressToWritable[address];
-                if (policy.IsTombstone)
+                EndpointManager.EndpointPolicy policy;
+                _addressToWritable.TryGetValue(address, out policy);
```
Comment (Aaronontheweb) [inline src/core/Akka.Remote/EndpointRegistry.cs]: Needed to do safe null-checking here.

File: src/core/Akka.Remote/EndpointRegistry.cs [hunk 8]
```diff
@@ -144,7 +146,8 @@ public EndpointManager.EndpointPolicy WritableEndpointWithPolicyFor(Address addr
 
         public bool HasWriteableEndpointFor(Address address)
         {
-            return WritableEndpointWithPolicyFor(address) != null;
+            var policy = WritableEndpointWithPolicyFor(address);
+            return policy is EndpointManager.Pass;
```
Comment (Aaronontheweb) [inline src/core/Akka.Remote/EndpointRegistry.cs]: This was a bad bug - previously we're return true, that a node is writeable as long as it had ANY policy. Not true - needs to be `Pass` in order to be writeable.

File: src/core/Akka.Remote/EndpointRegistry.cs [hunk 9]
```diff
@@ -155,7 +158,7 @@ public void MarkAsFailed(IActorRef endpoint, Deadline timeOfRelease)
         {
             if (IsWritable(endpoint))
             {
-                _addressToWritable.AddOrSet(_writableToAddress[endpoint], new EndpointManager.Gated(timeOfRelease));
+                _addressToWritable[_writableToAddress[endpoint]] = new EndpointManager.Gated(timeOfRelease);
```
Comment (Aaronontheweb) [inline src/core/Akka.Remote/EndpointRegistry.cs]: Replacing `AddOrSet` with the built-in operators.

File: src/core/Akka.Remote/EndpointRegistry.cs [hunk 10]
```diff
@@ -167,7 +170,7 @@ public void MarkAsFailed(IActorRef endpoint, Deadline timeOfRelease)
 
         public void MarkAsQuarantined(Address address, int uid, Deadline timeOfRelease)
         {
-            _addressToWritable.AddOrSet(address, new EndpointManager.Quarantined(uid, timeOfRelease));
+            _addressToWritable[address] = new EndpointManager.Quarantined(uid, timeOfRelease);
```
Comment (Aaronontheweb) [inline src/core/Akka.Remote/EndpointRegistry.cs]: Replacing `AddOrSet` with the built-in operators.

**Comment by Aaronontheweb** [comment]
This one is ready to review cc @annymsMthd

**Comment by Aaronontheweb** [comment]
Hrmmm.... hold the phone - found another potential issue that could be part of #1700

**Comment by Aaronontheweb** [comment]
Ok, did a pretty full review of the diff on #1755 - only thing I could come up with was that we had two bugs that offset each other and I only fixed one of them in 1.0.6. One bug had us pass in the wrong UID to the ProtocolStateActor all the way down from the `EndpointManager` and the second bug was falsely reporting that we needed a read-only endpoint, which resulted in the `ReliableDeliverySupervisor` not being started correctly.

This PR fixes the second bug, which was caused by an issue with the `EndpointRegistry.` Looking at including some more MNTR specs along with this...

**Comment by maxim-s** [comment]
Great work

**Pull request merged**